### PR TITLE
Remove chain id from Client interface

### PR DESCRIPTION
**Pull request by war-in**
### **Please check if the PR fulfills these requirements**
- [ ] The formatter, linter, and tests all run without an error
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

## **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

It removes unnecessary chain parameter from the Client interface

## **What is the current behaviour?** (You can also link to an open issue here)

Chain id is used only in the AccountClient to sign transactions, but it is in the Client interface

## **What is the new behaviour (if this is a feature change)?**

Chain id is present only in the AccontClient

## **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

Yes, the Client interface has changed

### CODE REVIEW

File: starknet_py/net/account/account_client.py [hunk 1]
```diff
@@ -55,34 +55,39 @@ def __init__(
...
+            chain = chain_from_network(net=client.net, chain=chain)
+            signer = StarkCurveSigner(
+                account_address=self.address, key_pair=key_pair, chain_id=chain
+            )
+        self.signer = signer
 
     @property
     def chain(self) -> StarknetChainId:
-        return self.client.chain
+        return self.signer.chain_id
```
Comment (cptartur) [inline starknet_py/net/account/account_client.py:90]: I wonder if it makes sense to expose this property here at all if it's going through signer anyway.

File: starknet_py/net/account/account_client.py [hunk 2]
```diff
@@ -55,34 +55,39 @@ def __init__(
...
-        if chain is None and client is None:
-            raise ValueError("One of chain or client must be provided")
         self.address = parse_address(address)
         self.client = client
-        self.signer = signer or StarkCurveSigner(
-            account_address=self.address, key_pair=key_pair, chain_id=self.client.chain
-        )
+
+        if signer is None:
+            chain = chain_from_network(net=client.net, chain=chain)
```
Comment (cptartur) [inline starknet_py/net/account/account_client.py:82]: What if someone has provided a chain? This way we're overwriting it. Shouldn't it be something along the lines of `chain = chain or ...`

File: starknet_py/net/account/account_client.py [hunk 3]
```diff
@@ -370,14 +376,15 @@ async def create_account(
         :param private_key: Private Key used for the account
         :param signer: Signer used to create account and sign transaction
         :return: Instance of AccountClient which interacts with created account on given network
+        :param chain: ChainId of the chain used by the client
```
Comment (cptartur) [inline starknet_py/net/account/account_client.py:379]: We should be doing some verification here if either signer or chain was provided, same as in constructor.

File: docs/contract.rst [hunk 4]
```diff
@@ -12,7 +12,7 @@ Contract
 
 .. autoclass:: PreparedFunctionCall
     :exclude-members: __init__, __new__
-    :members: call, call_raw, invoke, call_sync, call_raw_sync, invoke_sync, hash, estimate_fee, estimate_fee_sync
```
Comment (cptartur) [inline docs/contract.rst:15]: This removal needs to be mentioned in the release notes of the next version.

**Review by cptartur — CHANGES_REQUESTED** [review]

File: starknet_py/net/account/account_client.py (hunk 2 above)
Comment (war-in) [reply starknet_py/net/account/account_client.py:82]: But the chain is overridden only if net=("mainnet" or "testnet"). Isn't it ok? What should we do if someone provides the wrong chain while using these two networks?

**Review by cptartur — APPROVED** [review]

File: starknet_py/contract_test.py [hunk 5]
```diff
@@ -119,24 +118,6 @@ def test_compute_address_throws_on_no_source():
     )
 
 
-def test_transaction_hash():
```
Comment (Arcticae) [inline starknet_py/contract_test.py:122]: So there's no possibility to compute tx hash beforehand right now?

**Review by Arcticae — APPROVED** [review]

File: starknet_py/contract_test.py (hunk 5 above)
Comment (war-in) [reply starknet_py/contract_test.py:122]: There is an option. Invoke function has the "hash" property so it is possible

**Comment by war-in** [comment]
It will be merged someday, but first, we must deprecate the chain.

**Pull request closed without merging**